### PR TITLE
fix: unable to connect cluster when AutomountServiceAccountToken is disabled. Fixes #10937

### LINKED ISSUE #10937: Unable to connect cluster when specified executor serviceAccountName
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I can confirm the issues exists when I tested with `:latest`
- [X] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/master/docs/CONTRIBUTING.md))

### What happened/what you expected to happen?

**What happened**

Occurred below errors while creating resource via container/script template

```
create-resources-via-container-template: error: failed to create serviceaccount: Post "http://localhost:8080/api/v1/namespaces/default/serviceaccounts?fieldManager=kubectl-create&fieldValidation=Strict": dia
l tcp 127.0.0.1:8080: connect: connection refused
```

**what you expected to happen?**

Able to create resource via container/script template successfully.

### Version

latest

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
metadata:
  name: create-resources-via-container-template
  namespace: argo
spec:
  executor:
    serviceAccountName: sa-writer
  serviceAccountName: sa-writer
  automountServiceAccountToken: false
  entrypoint: main
  templates:
    - name: main
      container:
        name: main
        image: bitnami/kubectl
        command:
          - sh
        args:
          - -c
          - |
           kubectl create sa test
```

### Logs from the workflow controller

```text
healthy
```

### Logs from in your workflow's wait container

```text
healthy
```

### PR DESCRIPTION
**Pull request by maxsxu**
Fixes #10937

### CODE REVIEW

**Review by terrytangyuan — COMMENTED** [review]
Thanks. Could you add a test in https://github.com/argoproj/argo-workflows/tree/master/test/e2e?

**Comment by maxsxu** [comment]
> Thanks. Could you add a test in https://github.com/argoproj/argo-workflows/tree/master/test/e2e?

@terrytangyuan Added two e2e tests, PTAL

**Comment by terrytangyuan** [comment]
Test failure is related to your new test

**Comment by maxsxu** [comment]
> Test failure is related to your new test

Fixed by adding Secret/argo.service-account-token. Now all CI passed.

It's due to the KSA token secret issue of K8s 1.24+, the `CI / E2E Tests (test-functional, minimal) ` runs on 1.26 so it failed, but the `CI / E2E Tests (test-functional, v1.21.2+k3s1, minimal)` runs on 1.21 so it succeeds.

**Review by alexec — APPROVED** [review]

**Review by terrytangyuan — APPROVED** [review]

**Comment by agilgur5** [comment]
Per https://github.com/argoproj/argo-workflows/issues/12848#issuecomment-2026546244, this seems to have caused a regression: an SA token shouldn't be mounted if `automountServiceAccountToken: false` -- you have literally instructed it not to. An error trying to reach the k8s API with no SA token would be correct behavior.

File: workflow/controller/workflowpod.go [hunk 1]
```diff
@@ -222,6 +222,22 @@ func (woc *wfOperationCtx) createWorkflowPod(ctx context.Context, nodeName strin
...
+				continue
+			}
+			c.VolumeMounts = append(c.VolumeMounts, apiv1.VolumeMount{
+				Name:      common.ServiceAccountTokenVolumeName,
+				MountPath: common.ServiceAccountTokenMountPath,
+				ReadOnly:  true,
+			})
+			pod.Spec.Containers[i] = c
+		}
+	}
```
Comment (agilgur5) [inline workflow/controller/workflowpod.go:239]: The [`setupServiceAccount` function](https://github.com/argoproj/argo-workflows/blob/1017c1dbe2ccf3f0a59ae68f04d954dd92e69673/workflow/controller/workflowpod.go#L1134) would also be the proper place to make such changes

File: test/e2e/workflow_test.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+          - sh
+        source:
+          kubectl get cm 
+`).
+		When().
+		SubmitWorkflow().
+		WaitForWorkflow(fixtures.ToBeSucceeded, time.Minute*10).
+		Then().
+		ExpectWorkflow(func(t *testing.T, metadata *metav1.ObjectMeta, status *wfv1.WorkflowStatus) {
+			assert.Equal(t, wfv1.WorkflowSucceeded, status.Phase)
```
Comment (agilgur5) [inline test/e2e/workflow_test.go:84]: This actually should fail; the `main` container shouldn't have an accessible SA if you specified `automountServiceAccountToken: false` -- you have literally instructed it not to have one

**Pull request merged**